`allow_alias` enum option is not respected

https://github.com/googleapis/python-recommendations-ai/issues/14

> `IN_STOCK` and `STOCK_STATE_UNSPECIFIED` both have a value of 0, without specifying `allow_alias=True`. This is causing deployment errors with protobuf.
> 
> The workaround is to specify no binary in requirements.txt:
> 
> ```
> protobuf==3.13.0 --no-binary=protobuf
> ```
> 
> https://github.com/googleapis/python-recommendations-ai/blob/90a06d1edb5b72403c6e039f99c90d1bdeef3337/google/cloud/recommendationengine_v1beta1/types/catalog.py#L167
> 
> **Error**
> 
> ```
> TypeError: Couldn't build proto file into descriptor pool!
> Invalid proto descriptor for file "google/cloud/recommendationengine_v1beta1/types/catalog.proto":
> google.cloud.recommendationengine.v1beta1.ProductCatalogItem.StockState: 
> "google.cloud.recommendationengine.v1beta1.ProductCatalogItem.IN_STOCK" uses the same enum value as
> "google.cloud.recommendationengine.v1beta1.ProductCatalogItem.STOCK_STATE_UNSPECIFIED". If this is
> intended, set 'option allow_alias = true;' to the enum definition.
> ```
> 
> **Docs**
> https://developers.google.com/protocol-buffers/docs/proto3

Protos: https://github.com/googleapis/googleapis/blob/9db36e1646684fc12657d735d9a388df3d5316e4/google/cloud/recommendationengine/v1beta1/catalog.proto#L133-L150

I attempted to re-generate the recommendations ai library and the generated unit tests no longer pass. 

https://github.com/googleapis/python-recommendations-ai/pull/18

Lots of errors like this:

```
tests/unit/gapic/recommendationengine_v1beta1/test_user_event_service.py:39: in <module>
    from google.cloud.recommendationengine_v1beta1.services.user_event_service import (
google/cloud/recommendationengine_v1beta1/services/user_event_service/__init__.py:18: in <module>
    from .client import UserEventServiceClient
google/cloud/recommendationengine_v1beta1/services/user_event_service/client.py:38: in <module>
    from google.cloud.recommendationengine_v1beta1.services.user_event_service import pagers
google/cloud/recommendationengine_v1beta1/services/user_event_service/pagers.py:20: in <module>
    from google.cloud.recommendationengine_v1beta1.types import user_event
google/cloud/recommendationengine_v1beta1/__init__.py:18: in <module>
    from .services.catalog_service import CatalogServiceClient
google/cloud/recommendationengine_v1beta1/services/catalog_service/__init__.py:18: in <module>
    from .client import CatalogServiceClient
google/cloud/recommendationengine_v1beta1/services/catalog_service/client.py:37: in <module>
    from google.cloud.recommendationengine_v1beta1.services.catalog_service import pagers
google/cloud/recommendationengine_v1beta1/services/catalog_service/pagers.py:20: in <module>
    from google.cloud.recommendationengine_v1beta1.types import catalog
google/cloud/recommendationengine_v1beta1/types/__init__.py:19: in <module>
    from .catalog import (
google/cloud/recommendationengine_v1beta1/types/catalog.py:236: in <module>
    class Image(proto.Message):
.nox/unit-3-8/lib/python3.8/site-packages/proto/message.py:263: in __new__
    file_info.generate_file_pb(new_class=cls, fallback_salt=full_name)
.nox/unit-3-8/lib/python3.8/site-packages/proto/_file_info.py:102: in generate_file_pb
    pool.Add(self.descriptor)
E   TypeError: Couldn't build proto file into descriptor pool!
E   Invalid proto descriptor for file "google/cloud/recommendationengine_v1beta1/types/catalog.proto":
E     google.cloud.recommendationengine.v1beta1.CatalogItem.recommendation_type: "recommendation_type" is already defined in "google.cloud.recommendationengine.v1beta1.CatalogItem".
E     google.cloud.recommendationengine.v1beta1.CatalogItem.id: "id" is already defined in "google.cloud.recommendationengine.v1beta1.CatalogItem".
E     google.cloud.recommendationengine.v1beta1.CatalogItem.category_hierarchies: "category_hierarchies" is already defined in "google.cloud.recommendationengine.v1beta1.CatalogItem".
E     google.cloud.recommendationengine.v1beta1.CatalogItem.title: "title" is already defined in "google.cloud.recommendationengine.v1beta1.CatalogItem".
E     google.cloud.recommendationengine.v1beta1.CatalogItem.description: "description" is already defined in "google.cloud.recommendationengine.v1beta1.CatalogItem".
E     google.cloud.recommendationengine.v1beta1.CatalogItem.item_attributes: "item_attributes" is already defined in "google.cloud.recommendationengine.v1beta1.CatalogItem".
E     google.cloud.recommendationengine.v1beta1.CatalogItem.language_code: "language_code" is already defined in "google.cloud.recommendationengine.v1beta1.CatalogItem".
E     google.cloud.recommendationengine.v1beta1.CatalogItem.tags: "tags" is already defined in "google.cloud.recommendationengine.v1beta1.CatalogItem".
E     google.cloud.recommendationengine.v1beta1.CatalogItem.item_group_id: "item_group_id" is already defined in "google.cloud.recommendationengine.v1beta1.CatalogItem".
E     google.cloud.recommendationengine.v1beta1.CatalogItem.product_metadata: "product_metadata" is already defined in "google.cloud.recommendationengine.v1beta1.CatalogItem".
E     google.cloud.recommendationengine.v1beta1.CatalogItem.CategoryHierarchy.categories: "categories" is already defined in "google.cloud.recommendationengine.v1beta1.CatalogItem.CategoryHierarchy".
E     google.cloud.recommendationengine.v1beta1.CatalogItem.CategoryHierarchy: "CategoryHierarchy" is already defined in "google.cloud.recommendationengine.v1beta1.CatalogItem".
```

--- a/.github/sync-repo-settings.yaml
+++ b/.github/sync-repo-settings.yaml
@@ -12,7 +12,7 @@ branchProtectionRules:
     - 'unit (3.7)'
     - 'unit (3.7, cpp)'
     - 'unit (3.8)'
-    - 'unit (3.9, cpp)'
+    # - 'unit (3.9, cpp)' # Don't have binary wheels for 3.9 cpp protobuf yet
     - 'unit (3.9)'
     - 'cla/google'
   requiredApprovingReviewCount: 1

--- a/proto/enums.py
+++ b/proto/enums.py
@@ -24,6 +24,14 @@ from proto.marshal.rules.enums import EnumRule
 class ProtoEnumMeta(enum.EnumMeta):
     """A metaclass for building and registering protobuf enums."""
 
+    @classmethod
+    def __prepare__(mcls, cls, bases):
+        # The _ignore_ attribute must be set before any of the enum attributes
+        # are added because _ignore_ cannot specify already set names.
+        enum_dict = super().__prepare__(cls, bases)
+        enum_dict["_ignore_"] = ["_pb_options"]
+        return enum_dict
+
     def __new__(mcls, name, bases, attrs):
         # Do not do any special behavior for `proto.Enum` itself.
         if bases[0] == enum.IntEnum:
@@ -47,6 +55,13 @@ class ProtoEnumMeta(enum.EnumMeta):
         filename = _file_info._FileInfo.proto_file_name(
             attrs.get("__module__", name.lower())
         )
+
+        # Retrieve any enum options.
+        # We expect something that looks like an EnumOptions message,
+        # either an actual instance or a dict-like representation.
+        opts = attrs.pop("_pb_options", {})
+
+        # Make the descriptor.
         enum_desc = descriptor_pb2.EnumDescriptorProto(
             name=name,
             # Note: the superclass ctor removes the variants, so get them now.
@@ -60,6 +75,7 @@ class ProtoEnumMeta(enum.EnumMeta):
                 ),
                 key=lambda v: v.number,
             ),
+            options=opts,
         )
 
         file_info = _file_info._FileInfo.maybe_add_descriptor(filename, package)

--- a/proto/enums.py
+++ b/proto/enums.py
@@ -24,14 +24,6 @@ from proto.marshal.rules.enums import EnumRule
 class ProtoEnumMeta(enum.EnumMeta):
     """A metaclass for building and registering protobuf enums."""
 
-    @classmethod
-    def __prepare__(mcls, cls, bases):
-        # The _ignore_ attribute must be set before any of the enum attributes
-        # are added because _ignore_ cannot specify already set names.
-        enum_dict = super().__prepare__(cls, bases)
-        enum_dict["_ignore_"] = ["_pb_options"]
-        return enum_dict
-
     def __new__(mcls, name, bases, attrs):
         # Do not do any special behavior for `proto.Enum` itself.
         if bases[0] == enum.IntEnum:
@@ -59,7 +51,15 @@ class ProtoEnumMeta(enum.EnumMeta):
         # Retrieve any enum options.
         # We expect something that looks like an EnumOptions message,
         # either an actual instance or a dict-like representation.
-        opts = attrs.pop("_pb_options", {})
+        pb_options = "_pb_options"
+        opts = attrs.pop(pb_options, {})
+        # This is the only portable way to remove the _pb_options name
+        # from the enum attrs.
+        # In 3.7 onwards, we can define an _ignore_ attribute and do some
+        # mucking around with that.
+        if pb_options in attrs._member_names:
+            idx = attrs._member_names.index(pb_options)
+            attrs._member_names.pop(idx)
 
         # Make the descriptor.
         enum_desc = descriptor_pb2.EnumDescriptorProto(

--- a/tests/test_fields_enum.py
+++ b/tests/test_fields_enum.py
@@ -12,7 +12,9 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import os
 import proto
+import pytest
 import sys
 
 
@@ -353,3 +355,40 @@ def test_unwrapped_enum_fields():
     t = Task(weekday="TUESDAY")
     t2 = Task.deserialize(Task.serialize(t))
     assert t == t2
+
+
+if os.environ.get("PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION", "python") == "cpp":
+    # This test only works, and is only relevant, with the cpp runtime.
+    # Python just doesn't give a care and lets it work anyway.
+    def test_enum_alias_bad():
+        # Certain enums may shadow the different enum monikers with the same value.
+        # This is generally discouraged, and protobuf will object by default,
+        # but will explicitly allow this behavior if the enum is defined with
+        # the `allow_alias` option set.
+        with pytest.raises(TypeError):
+
+            # The wrapper message is a hack to avoid manifest wrangling to
+            # define the enum.
+            class BadMessage(proto.Message):
+                class BadEnum(proto.Enum):
+                    UNKNOWN = 0
+                    DEFAULT = 0
+
+                bad_dup_enum = proto.Field(proto.ENUM, number=1, enum=BadEnum)
+
+
+def test_enum_alias_good():
+    # Have to split good and bad enum alias into two tests so that the generated
+    # file descriptor is properly created.
+    # For the python runtime, aliases are allowed by default, but we want to
+    # make sure that the options don't cause problems.
+    # For the cpp runtime, we need to verify that we can in fact define aliases.
+    class GoodMessage(proto.Message):
+        class GoodEnum(proto.Enum):
+            _pb_options = {"allow_alias": True}
+            UNKNOWN = 0
+            DEFAULT = 0
+
+        good_dup_enum = proto.Field(proto.ENUM, number=1, enum=GoodEnum)
+
+    assert GoodMessage.GoodEnum.UNKNOWN == GoodMessage.GoodEnum.DEFAULT == 0

